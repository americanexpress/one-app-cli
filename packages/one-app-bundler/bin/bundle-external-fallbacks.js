const path = require('path');
const webpack = require('webpack');
const readPkgUp = require('read-pkg-up');
const chalk = require('chalk');
const { ConcatSource } = require('webpack-sources');
const ModuleFilenameHelpers = require('webpack/lib/ModuleFilenameHelpers');
const getRegisterExternalStr = require('../utils/getRegisterExternalStr');
const getExternalLibraryName = require('../utils/getExternalLibraryName');

function HolocronExternalRegisterPlugin(externalName, version) {
  this.externalName = externalName;
  this.externalVersion = version;
  this.options = {};
}

HolocronExternalRegisterPlugin.prototype.apply = function apply(compiler) {
  const { externalName, externalVersion, options } = this;
  compiler.hooks.compilation.tap('HolocronExternalRegisterPlugin', (compilation) => {
    compilation.hooks.optimizeChunkAssets.tapAsync('HolocronExternalRegisterPlugin', (chunks, callback) => {
      chunks.forEach((chunk) => {
        if (chunk.name !== 'main') return;
        chunk.files
          .filter(ModuleFilenameHelpers.matchObject.bind(undefined, options))
          .forEach((file) => {
            // eslint-disable-next-line no-param-reassign -- this is a webpack mutation
            compilation.assets[file] = new ConcatSource(
              compilation.assets[file],
              '\n',
              getRegisterExternalStr(externalName, externalVersion),
              '\n'
            );
          });
      });

      callback();
    });
  });
};

module.exports = function bundleExternalFallbacks() {
  const { packageJson } = readPkgUp.sync();
  const { 'one-amex': { bundler = {} } } = packageJson;
  const { requiredExternals } = bundler;

  if (
    !requiredExternals || !Array.isArray(requiredExternals)
    || requiredExternals.length === 0
  ) {
    return;
  }

  requiredExternals.forEach((externalName) => {
    const indexPath = path.resolve(process.cwd(), `node_modules/${externalName}`);
    // eslint-disable-next-line global-require, import/no-dynamic-require -- need to require a package.json at runtime
    const { version } = require(`${externalName}/package.json`);

    webpack({
      entry: indexPath,
      output: {
        path: path.resolve(process.cwd(), `build/${packageJson.version}`),
        filename: `${externalName}.js`,
        library: `${getExternalLibraryName(externalName, version)}`,
      },
      plugins: [
        new HolocronExternalRegisterPlugin(externalName, version),
      ],
    }, (externalError) => {
      if (externalError) {
        console.log(`Failed to bundle external - ${externalName}`);
        console.log(chalk.red(externalError), chalk.red(externalError.stack));
        throw externalError;
      }
    });
  });
};
