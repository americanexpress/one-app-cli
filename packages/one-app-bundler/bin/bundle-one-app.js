#! /usr/bin/env node
/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const chalk = require('chalk');

const postProcessOneAppBundle = require('./postProcessOneAppBundle');
const createClientConfig = require('../webpack/app/webpack.client');
const buildWebpack = require('../utils/buildWebpack');
const time = require('../utils/time');

(async function bundleOneApp() {
  await (
    time(() => {
      const configs = [
        ['browser', createClientConfig('modern')],
        ['legacyBrowser', createClientConfig('legacy')],
      ].map(([name, config]) => ({
        ...config,
        name,
      }));

      return buildWebpack(configs)
        .then(postProcessOneAppBundle)
        .catch((err) => {
          console.log(chalk.red(err), chalk.red(err.stack));
        });
    })
      .then((timeInMs) => {
        console.log(timeInMs);
      })
  );
}());
