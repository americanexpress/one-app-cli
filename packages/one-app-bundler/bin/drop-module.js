#! /usr/bin/env node
/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const fs = require('fs');
const path = require('path');
const { argv } = require('yargs');

const publicPath = path.join(process.cwd(), 'static');
const symModulesPath = path.join(publicPath, 'modules');
const moduleMapPath = path.join(publicPath, 'module-map.json');

argv._.forEach((moduleName) => {
  try {
    fs.accessSync(moduleMapPath);
  } catch (e) {
    throw new Error('No modules are currently being served');
  }

  const moduleMap = JSON.parse(fs.readFileSync(moduleMapPath));
  delete moduleMap.modules[moduleName];
  fs.writeFileSync(moduleMapPath, JSON.stringify(moduleMap));

  const modulePath = path.join(symModulesPath, moduleName);

  try {
    fs.rmdirSync(modulePath);
  } catch (e) {
    // Do nothing, file does not exist, so it does not need to be deleted
  }
});
