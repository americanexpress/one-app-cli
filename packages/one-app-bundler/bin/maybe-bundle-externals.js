const path = require('path');
const webpack = require('webpack');
const readPkgUp = require('read-pkg-up');
const chalk = require('chalk');
const { snakeCase } = require('lodash');
const { ConcatSource } = require('webpack-sources');
const ModuleFilenameHelpers = require('webpack/lib/ModuleFilenameHelpers');
const { EXTERNAL_PREFIX } = require('..');
function HolocronModuleRegisterPlugin(externalName) {
  this.externalName = externalName;
  this.options = {};
}

HolocronModuleRegisterPlugin.prototype.apply = function apply(compiler) {
  const { externalName, options } = this;
  compiler.hooks.compilation.tap('HolocronModuleRegisterPlugin', (compilation) => {
    compilation.hooks.optimizeChunkAssets.tapAsync('HolocronModuleRegisterPlugin', (chunks, callback) => {
      chunks.forEach((chunk) => {
        if (chunk.name !== 'main') return;
        chunk.files
          .filter(ModuleFilenameHelpers.matchObject.bind(undefined, options))
          .forEach((file) => {
            // eslint-disable-next-line no-param-reassign
            compilation.assets[file] = new ConcatSource(
              '(function() {',
              '\n',
              compilation.assets[file],
              '\n',
              `Holocron.registerExternal("${externalName}");})();`
            );
          });
      });

      callback();
    });
  });
};

module.exports = function maybeBundleExternals(runtimeEnv) {
  const { packageJson } = readPkgUp.sync();
  const { version, 'one-amex': { bundler } } = packageJson;
  const { requiredExternals } = bundler;

  if (
    !requiredExternals || !Array.isArray(requiredExternals)
    || requiredExternals.length === 0
  ) {
    return;
  }

  if (!['browser', 'node'].includes(runtimeEnv)) {
    throw new Error(`Invalid runtimeEnv "${runtimeEnv}"`);
  }

  requiredExternals.forEach((externalName) => {
    const indexPath = path.resolve(process.cwd(), `node_modules/${externalName}`);

    webpack({
      entry: indexPath,
      output: {
        path: path.resolve(process.cwd(), `build/${version}`),
        filename: `${externalName}.${runtimeEnv}.js`,
        ...runtimeEnv === 'browser' ? {
          library: `${EXTERNAL_PREFIX}${snakeCase(externalName)}`,
        } : {
          libraryTarget: 'commonjs2',
        },
      },
      plugins: [
        new HolocronModuleRegisterPlugin(externalName),
      ],
    }, (externalError) => {
      if (externalError) {
        console.log(`Failed to bundle external - ${externalName} (runtimeEnv)`);
        console.log(chalk.red(externalError), chalk.red(externalError.stack));
        throw externalError;
      }
    });
  });
};
