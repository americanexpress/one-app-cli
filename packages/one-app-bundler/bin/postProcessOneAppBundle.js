/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable import/no-dynamic-require, global-require */

const path = require('path');
const fs = require('fs');
const { hashElement } = require('folder-hash');
const readPkgUp = require('read-pkg-up');
const generateIntegrityManifest = require('./generateIntegrityManifest');

const { packageJson, path: pkgPath } = readPkgUp.sync();
const { version } = packageJson;
const tmpPath = path.resolve(pkgPath, '../build/app/tmp');

module.exports = async function postProcessBuild() {
  const endsWithJS = (fileName) => fileName.endsWith('.js');
  const legacyPath = path.join(tmpPath, 'legacy');
  const jsFileNames = fs.readdirSync(tmpPath).filter(endsWithJS);
  const legacyJsFileNames = fs.readdirSync(legacyPath).filter(endsWithJS);
  const addIntegrityToManifest = (pathName, prefix = '') => (fileName) => generateIntegrityManifest(prefix + fileName, path.join(pathName, fileName));
  jsFileNames.forEach(addIntegrityToManifest(tmpPath));
  legacyJsFileNames.forEach(addIntegrityToManifest(legacyPath, 'legacy/'));
  const options = {
    files: { include: ['*.js'] },
    encoding: 'hex',
  };
  const { hash } = await hashElement(tmpPath, options);
  const buildVersion = `${version}-${hash.substring(0, 8)}`;
  const modernBrowserChunkAssets = require(path.resolve(pkgPath, '../.webpack-stats.browser.json')).assetsByChunkName;
  const legacyBrowserChunkAssets = require(path.resolve(pkgPath, '../.webpack-stats.legacyBrowser.json')).assetsByChunkName;

  fs.renameSync(tmpPath, path.resolve(tmpPath, `../${buildVersion}`));
  const metaFilePath = path.resolve(pkgPath, '../.build-meta.json');
  if (fs.existsSync(metaFilePath)) fs.unlinkSync(metaFilePath);
  fs.writeFileSync(
    metaFilePath,
    JSON.stringify(
      { buildVersion, modernBrowserChunkAssets, legacyBrowserChunkAssets },
      undefined,
      2
    )
  );
};
