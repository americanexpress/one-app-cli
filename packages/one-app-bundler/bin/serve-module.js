#! /usr/bin/env node
/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const fs = require('fs');
const path = require('path');
const mkdirp = require('mkdirp');
const { argv } = require('yargs');
const rimraf = require('rimraf');

const publicPath = path.join(process.cwd(), 'static');
const symModulesPath = path.join(publicPath, 'modules');

mkdirp.sync(symModulesPath);

argv._.forEach((modulePath) => {
  const pkg = JSON.parse(fs.readFileSync(path.join(modulePath, 'package.json')));
  const absoluteModulePath = modulePath.startsWith('/')
    ? modulePath : path.join(process.cwd(), modulePath);
  const {
    browser: browserSri,
    legacyBrowser: legacyBrowserSri,
    node: nodeSri,
  } = JSON.parse(fs.readFileSync(path.join(modulePath, 'bundle.integrity.manifest.json')));
  const { version, name: moduleName } = pkg;

  if (!version) {
    throw new Error(`No versioned bundles exist for module ${moduleName}`);
  }

  // enforce only one version of a module being served at once
  rimraf.sync(path.join(symModulesPath, moduleName));
  mkdirp.sync(path.join(symModulesPath, moduleName));

  const sourceBundlePath = path.join(absoluteModulePath, 'build', version);
  const symBundlePath = path.join(symModulesPath, moduleName, version);

  try {
    fs.accessSync(symBundlePath);
    fs.unlinkSync(symBundlePath);
  } catch (e) {
    // Do nothing, file does not exist, so it does not need to be deleted
  } finally {
    /*
     * Cant symlink on windows without admin, windows uses junctions
     * we only symlink dir so file type not needed
     */
    const type = process.platform === 'win32' ? 'junction' : 'dir';
    fs.symlinkSync(sourceBundlePath, symBundlePath, type);
  }

  const moduleMapPath = path.join(publicPath, 'module-map.json');
  try {
    fs.accessSync(moduleMapPath);
  } catch (e) {
    fs.writeFileSync(moduleMapPath, JSON.stringify({ key: 'not-used-in-development', modules: {} }, null, 2));
  } finally {
    const moduleMap = JSON.parse(fs.readFileSync(moduleMapPath));
    moduleMap.modules[moduleName] = {
      browser: {
        integrity: browserSri,
        url: `[one-app-dev-cdn-url]/static/modules/${moduleName}/${version}/${moduleName}.browser.js`,
      },
      legacyBrowser: {
        integrity: legacyBrowserSri,
        url: `[one-app-dev-cdn-url]/static/modules/${moduleName}/${version}/${moduleName}.legacy.browser.js`,
      },
      node: {
        integrity: nodeSri,
        url: `[one-app-dev-cdn-url]/static/modules/${moduleName}/${version}/${moduleName}.node.js`,
      },
    };

    fs.writeFileSync(moduleMapPath, JSON.stringify(moduleMap, null, 2));
  }
});
