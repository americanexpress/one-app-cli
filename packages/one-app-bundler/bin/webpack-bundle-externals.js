const path = require('path');
const webpack = require('webpack');
const readPkgUp = require('read-pkg-up');
const chalk = require('chalk');
const { snakeCase } = require('lodash');
const { EXTERNAL_PREFIX } = require('..');

module.exports = function bundleExternals(label) {
  const { packageJson } = readPkgUp.sync();
  const { version } = packageJson;
  const { bundler } = packageJson['one-amex'];

  if (!bundler.sharedExternals || bundler.sharedExternals.length === 0) {
    return;
  }

  if (label === 'browser') {
    bundler.sharedExternals.forEach((external) => {
      webpack({
        entry: path.resolve(process.cwd(), `node_modules/${external}`),
        output: {
          // This makes the global variable name `__holocron_external_<package-name>`
          // eg. `__holocron_external_is_even`
          library: `${EXTERNAL_PREFIX}${snakeCase(external)}`,
          path: path.resolve(process.cwd(), `build/${version}`),
          filename: `${external}.browser.js`,
        },
      }, (externalError) => {
        if (externalError) {
          console.log(`Failed to bundle external - ${external} (browser)`);
          console.log(chalk.red(externalError), chalk.red(externalError.stack));
          throw externalError;
        }
      });
    });
  }
  if (label === 'node') {
    bundler.sharedExternals.forEach((external) => {
      webpack({
        entry: path.resolve(process.cwd(), `node_modules/${external}`),
        output: {
          path: path.resolve(process.cwd(), `build/${version}`),
          filename: `${external}.node.js`,
          libraryTarget: 'commonjs2',
        },
      }, (externalError) => {
        if (externalError) {
          console.log(`Failed to bundle external - ${external} (node)`);
          console.log(chalk.red(externalError), chalk.red(externalError.stack));
          throw externalError;
        }
      });
    });
  }
};
