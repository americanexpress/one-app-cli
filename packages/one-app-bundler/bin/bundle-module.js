#! /usr/bin/env node
/*
 * Copyright 2022 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */
import path from 'node:path';
import fs from 'node:fs';

import { bundleExternalFallbacks } from './bundle-external-fallbacks.js';

const bundleModule = async () => {

  fs.writeFileSync(path.join(process.cwd(), 'bundle.integrity.manifest.json'), JSON.stringify({}));

  await bundleExternalFallbacks();

  if (process.env.NODE_ENV !== 'production' && process.argv.includes('--dev')) {
    console.info('Running dev bundler');

    const {
      devBuildModule,
    } = await import('@americanexpress/one-app-dev-bundler');

    if (devBuildModule) {
      devBuildModule().catch((error) => {
        console.error(`Build failed with error ${error.message}`);
        console.error(error);
        throw error;
      });
    }
  } else {
    if (process.argv.includes('--dev')) {
      console.info('Ignoring `--dev` flag for NODE_ENV=production');
    }
    console.info('Running production bundler');
    const { webpackBundleModule } = await import('./webpack-bundle-module.js');
    await webpackBundleModule();
  }
};

bundleModule();
