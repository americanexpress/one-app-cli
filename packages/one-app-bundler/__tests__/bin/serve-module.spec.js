/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable global-require */
let fs = require('fs');

let rimraf;

jest.mock('fs');

const setup = (modulePath) => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.doMock('yargs', () => ({
    argv: { _: [modulePath] },
  }));
  jest.doMock('rimraf', () => ({
    sync: jest.fn(),
  }));

  fs = require('fs');
  rimraf = require('rimraf');
};

describe('serve-module', () => {
  process.cwd = () => '/mocked';
  const originalPlatform = process.platform;

  const setPlatform = (platform) => Object.defineProperty(process, 'platform', {
    value: platform,
  });

  beforeEach(() => {
    setup('../my-module-name');
  });

  afterAll(() => {
    setPlatform(originalPlatform);
  });

  it('should throw if the module doesn\'t have a version', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name' }),
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    expect(() => require('../../bin/serve-module')).toThrowErrorMatchingSnapshot();
  });

  it('should create a directory for the module', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name', version: '1.0.0' }),
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    require('../../bin/serve-module');
    expect(fs.mkdirSync).toHaveBeenCalledTimes(2);
    expect(fs.mkdirSync.mock.calls[0][0]).toEqual('/mocked/static/modules');
    expect(fs.mkdirSync.mock.calls[1][0]).toEqual('/mocked/static/modules/my-module-name');
  });

  it('should remove a directory for the module if it already exists since only one version of a module can be served at a time', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name', version: '1.0.0' }),
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
      '/mocked/static/modules/my-module-name/1.0.0': {},
    });
    require('../../bin/serve-module');
    expect(rimraf.sync).toHaveBeenCalledWith('/mocked/static/modules/my-module-name');
  });

  it('should remove an existing symlink and create a new one', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name', version: '1.0.0' }),
      '/mocked/static/modules/my-module-name/1.0.0': {},
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    setPlatform('darwin');
    require('../../bin/serve-module');
    expect(fs.unlinkSync).toHaveBeenCalledTimes(1);
    expect(fs.unlinkSync).toHaveBeenCalledWith('/mocked/static/modules/my-module-name/1.0.0');
    expect(fs.symlinkSync).toHaveBeenCalledTimes(1);
    expect(fs.symlinkSync).toHaveBeenCalledWith('/my-module-name/build/1.0.0', '/mocked/static/modules/my-module-name/1.0.0', 'dir');
  });

  it('should work with an absolute module path', () => {
    setup('/absolute/path/to/module');
    fs._.setFiles({
      '/absolute/path/to/module/package.json': JSON.stringify({ name: 'absolute-module', version: '0.3.0' }),
      '/absolute/path/to/module/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    setPlatform('darwin');
    require('../../bin/serve-module');
    expect(fs.symlinkSync).toHaveBeenCalledWith('/absolute/path/to/module/build/0.3.0', '/mocked/static/modules/absolute-module/0.3.0', 'dir');
  });

  it('should use a junction on windows', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name', version: '1.0.0' }),
      '/mocked/static/modules/my-module-name/1.0.0': {
        'my-module-name.client.js': function MyModule() {}.toString(),
        'my-module-name.server.js': function MyModule() {}.toString(),
      },
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    const { platform } = process;
    setPlatform('win32');
    require('../../bin/serve-module');
    expect(fs.symlinkSync).toHaveBeenCalledWith('/my-module-name/build/1.0.0', '/mocked/static/modules/my-module-name/1.0.0', 'junction');
    process.platform = platform;
  });

  it('adds to the existing module map', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name', version: '1.0.0' }),
      '/mocked/static/module-map.json': JSON.stringify({
        key: '--- omitted for development ---',
        modules: {
          'another-module': {
            node: {
              url: 'https://example.com/cdn/another-module/6.7.8/another-module.node.js',
              integrity: '123',
            },
            browser: {
              url: 'https://example.com/cdn/another-module/6.7.8/another-module.browser.js',
              integrity: '234',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/another-module/6.7.8/another-module.legacy.browser.js',
              integrity: '345',
            },
          },
        },
      }),
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    require('../../bin/serve-module');
    expect(fs._.getFiles()['/mocked/static/module-map.json']).toMatchSnapshot();
  });

  it('creates a module map when one does not exist', () => {
    fs._.setFiles({
      '../my-module-name/package.json': JSON.stringify({ name: 'my-module-name', version: '3.4.2' }),
      '../my-module-name/bundle.integrity.manifest.json': JSON.stringify({ node: '123', browser: '234', legacyBrowser: '974' }),
    });
    require('../../bin/serve-module');
    expect(fs._.getFiles()['/mocked/static/module-map.json']).toMatchSnapshot();
  });
});
