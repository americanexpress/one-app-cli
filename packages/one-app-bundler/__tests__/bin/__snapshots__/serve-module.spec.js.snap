// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serve-module adds to the existing module map 1`] = `
"{
  \\"key\\": \\"--- omitted for development ---\\",
  \\"modules\\": {
    \\"another-module\\": {
      \\"node\\": {
        \\"url\\": \\"https://example.com/cdn/another-module/6.7.8/another-module.node.js\\",
        \\"integrity\\": \\"123\\"
      },
      \\"browser\\": {
        \\"url\\": \\"https://example.com/cdn/another-module/6.7.8/another-module.browser.js\\",
        \\"integrity\\": \\"234\\"
      },
      \\"legacyBrowser\\": {
        \\"url\\": \\"https://example.com/cdn/another-module/6.7.8/another-module.legacy.browser.js\\",
        \\"integrity\\": \\"345\\"
      }
    },
    \\"my-module-name\\": {
      \\"browser\\": {
        \\"integrity\\": \\"234\\",
        \\"url\\": \\"[one-app-dev-cdn-url]/static/modules/my-module-name/1.0.0/my-module-name.browser.js\\"
      },
      \\"legacyBrowser\\": {
        \\"integrity\\": \\"974\\",
        \\"url\\": \\"[one-app-dev-cdn-url]/static/modules/my-module-name/1.0.0/my-module-name.legacy.browser.js\\"
      },
      \\"node\\": {
        \\"integrity\\": \\"123\\",
        \\"url\\": \\"[one-app-dev-cdn-url]/static/modules/my-module-name/1.0.0/my-module-name.node.js\\"
      }
    }
  }
}"
`;

exports[`serve-module creates a module map when one does not exist 1`] = `
"{
  \\"key\\": \\"not-used-in-development\\",
  \\"modules\\": {
    \\"my-module-name\\": {
      \\"browser\\": {
        \\"integrity\\": \\"234\\",
        \\"url\\": \\"[one-app-dev-cdn-url]/static/modules/my-module-name/3.4.2/my-module-name.browser.js\\"
      },
      \\"legacyBrowser\\": {
        \\"integrity\\": \\"974\\",
        \\"url\\": \\"[one-app-dev-cdn-url]/static/modules/my-module-name/3.4.2/my-module-name.legacy.browser.js\\"
      },
      \\"node\\": {
        \\"integrity\\": \\"123\\",
        \\"url\\": \\"[one-app-dev-cdn-url]/static/modules/my-module-name/3.4.2/my-module-name.node.js\\"
      }
    }
  }
}"
`;

exports[`serve-module should throw if the module doesn't have a version 1`] = `"No versioned bundles exist for module my-module-name"`;
